hedera-node/hedera-app/src/test/java/com/hedera/node/app/platform/event/EventMigrationTest.java

```diff
@@ -0,0 +1,94 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.platform.event;
+
+import com.hedera.node.app.service.mono.context.properties.SerializableSemVers;
+import com.swirlds.common.constructable.ConstructableRegistry;
+import com.swirlds.common.constructable.ConstructableRegistryException;
+import com.swirlds.common.crypto.CryptographyHolder;
+import com.swirlds.common.crypto.Hash;
+import com.swirlds.platform.recovery.internal.EventStreamSingleFileIterator;
+import com.swirlds.platform.system.StaticSoftwareVersion;
+import com.swirlds.platform.system.events.BaseEventHashedData;
+import java.io.File;
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Stream;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+
+public class EventMigrationTest {
+
+    @BeforeAll
+    public static void setUp() throws ConstructableRegistryException {
+        ConstructableRegistry.getInstance().registerConstructables("");
+        StaticSoftwareVersion.setSoftwareVersion(Set.of(SerializableSemVers.CLASS_ID));
+    }
+
+    /**
+     * Tests the migration of events as we are switching events to protobuf. The main thing we are testing is that the
+     * hashes of old events can still be calculated when the code changes. This is done by calculating the hashes of the
+     * events that are read and matching them to the parent descriptors inside the events. The parents of most events
+     * will be present in the file, except for a few events at the beginning of the file.
+     * <p>
+     * The file being read is from mainnet written by the SDK 0.46.3.
+     * <p>
+     * Even though this could be considered a platform test, it needs to be in the services module because the event
+     * contains a {@link com.hedera.node.app.service.mono.context.properties.SerializableSemVers} which is a services
+     * class
+     */
+    @Test
+    public void migration() throws URISyntaxException, IOException {
+        final Set<Hash> eventHashes = new HashSet<>();
+        final Set<Hash> parentHashes = new HashSet<>();
+        int numEvents = 0;
+
+        try (final EventStreamSingleFileIterator iterator = new EventStreamSingleFileIterator(
+                new File(this.getClass()
+                                .getClassLoader()
+                                .getResource("eventFiles/sdk0.46.3/2024-03-05T00_10_55.002129867Z.events")
+                                .toURI())
+                        .toPath(),
+                false)) {
+            while (iterator.hasNext()) {
+                final BaseEventHashedData hashedData = iterator.next().getBaseEventHashedData();
+                numEvents++;
+                CryptographyHolder.get().digestSync(hashedData);
```

Comment:
How is this test performing the migration between event serialization schemes?

Reply:
i explain everything in the javadoc. I have not changed the event serialization yet, I'm about to. the main thing this is testing is that the hash is maintained after migration

Reply:
I'm trying to make sure I understand how this tests the migration.  The events on disk are in the old format, they are read into memory.  When they are hashed, the hashing serializes them in the new format?    Since all but 9 events act as parents to other events in the file, it is sufficient to compare the newly hashed values (derived from new serialization) to the stored hashed values in the parent descriptors (derived from the old serialization)?

Reply:
right now, there is no new format yet, this is what I'm currently working on.
what you said is correct. Ideally, we would have events that are stored together with their hashes. but we don't store event hashes, but we do store parent descriptors which do contain hashes, so we use those to determine if the hash has diverged